core-bundle/tests/Functional/ServiceArgumentsTest.php

```diff
@@ -0,0 +1,184 @@
+                if ('_' === $serviceId[0]) {
+                    continue;
+                }
+
+                if (!isset($config['class'])) {
+                    continue;
+                }
+
+                if (!$container->has($serviceId)) {
+                    continue;
+                }
+
+                $this->doTestService($serviceId, $config, $container);
+            }
+        }
+    }
+
+    private function doTestService(string $serviceId, array $config, ContainerInterface $container): void
+    {
+        $ref = new \ReflectionClass($config['class']);
+
+        if (!$constructor = $ref->getConstructor()) {
+            $this->markTestSkipped(sprintf('Service %s does not have a constructor method', $serviceId));
+        }
+
+        $arguments = $config['arguments'] ?? [];
+
+        $this->assertGreaterThanOrEqual(
+            $this->countRequiredParameters($constructor),
+            $config['arguments'] ?? [],
+            sprintf('Service %s does not have the necessary amount of parameters.', $serviceId)
+        );
+
+        foreach ($constructor->getParameters() as $i => $parameter) {
+            if (!\array_key_exists($i, $arguments)) {
+                $this->assertTrue($parameter->isOptional(), sprintf('Missing argument %s on service ID "%s".', $i, $serviceId));
+
+                continue;
+            }
+
+            $argument = $arguments[$i];
+            $type = $parameter->getType();
+            $typeName = $type instanceof \ReflectionNamedType ? $type->getName() : null;
+
+            if (null === $argument) {
+                if (!$parameter->allowsNull()) {
+                    $this->addWarning(sprintf('Argument %s ($%s) of %s does not allow NULL, assuming this parameter is set at runtime.', $i, $parameter->getName(), $serviceId));
+                }
+
+                continue;
+            }
+
+            if ($argument instanceof TaggedValue) {
+                switch ($argument->getTag()) {
+                    case 'service_closure':
+                        $this->assertSame('Closure', $typeName, sprintf('Argument %s of %s should be \Closure but found %s.', $i, $serviceId, $typeName));
+                        break;
+
+                    case 'tagged_iterator':
+                        $this->assertSame('iterable', $typeName, sprintf('Argument %s of %s is not an iterable', $i, $serviceId));
+                        break;
+
+                    case 'tagged_locator':
+                    case 'service_locator':
+                        $this->assertSame(PsrContainerInterface::class, $typeName, sprintf('Argument %s of %s should be %s but found %s.', $i, $serviceId, PsrContainerInterface::class, $typeName));
+                        break;
+
+                    default:
+                        $this->addWarning(sprintf('Unknown tagged type "%s" for argument %s ($%s) of service %s.', $parameter->getType(), $i, $parameter->getName(), $serviceId));
```

Comment:
Same here?

Reply:
changed in 819155bcb858f754baaa48e342f34e39095ef9fc